### PR TITLE
feat: add prisma adapter

### PR DESCRIPTION
**Pull request by williamluke4**
### Current Structure
```
├── LICENSE
├── package.json
├── packages
│   ├── example
│   │   ├── index.js
│   │   └── package.json
│   └── prisma
│       ├── package.json
│       ├── prisma
│       │   └── schema.prisma
│       ├── README.md
│       ├── src
│       │   └── index.ts
│       ├── tsconfig.json
│       └── yarn.lock
└── README.md
```

### Breaking Schema Changes from the original Prisma adapter
```diff
model Account {
-  id                 Int       @default(autoincrement()) @id
+  id                 Int       @id @default(autoincrement())
-  compoundId         String    @unique @map(name: "compound_id")
-  userId             Int       @map(name: "user_id")
+  userId             Int
+  user               User      @relation(fields: [userId], references: [id])
-  providerType       String    @map(name: "provider_type")
+  providerType       String
-  providerId         String    @map(name: "provider_id")
+  providerId         String
-  providerAccountId  String    @map(name: "provider_account_id")
+  providerAccountId  String
-  refreshToken       String?   @map(name: "refresh_token")
+  refreshToken       String?
-  accessToken        String?   @map(name: "access_token")
+  accessToken        String?
-  accessTokenExpires DateTime? @map(name: "access_token_expires")
+  accessTokenExpires DateTime?
-  createdAt          DateTime  @default(now()) @map(name: "created_at")
+  createdAt          DateTime  @default(now())
-  updatedAt          DateTime  @default(now()) @map(name: "updated_at")
+  updatedAt          DateTime  @updatedAt

-  @@index([providerAccountId], name: "providerAccountId")
-  @@index([providerId], name: "providerId")
-  @@index([userId], name: "userId")
-  @@map(name: "accounts")
+  @@unique([providerId, providerAccountId])
 }
 
 model Session {
-  id           Int      @default(autoincrement()) @id
+  id           Int      @id @default(autoincrement())
-  userId       Int      @map(name: "user_id")
+  userId       Int
+  user         User     @relation(fields: [userId], references: [id])
   expires      DateTime
-  sessionToken String   @unique @map(name: "session_token")
+  sessionToken String   @unique
-  accessToken  String   @unique @map(name: "access_token")
+  accessToken  String   @unique
-  createdAt    DateTime @default(now()) @map(name: "created_at")
+  createdAt    DateTime @default(now())
-  updatedAt    DateTime @default(now()) @map(name: "updated_at")
+  updatedAt    DateTime @updatedAt
-
-  @@map(name: "sessions")
 }
 
 model User {
-  id            Int       @default(autoincrement()) @id
+  id            Int       @id @default(autoincrement())
   name          String?
   email         String?   @unique
-  emailVerified DateTime? @map(name: "email_verified")
+  emailVerified DateTime?
   image         String?
+  accounts      Account[]
+  sessions      Session[]
-  createdAt     DateTime  @default(now()) @map(name: "created_at")
+  createdAt     DateTime  @default(now())
-  updatedAt     DateTime  @default(now()) @map(name: "updated_at")
+  updatedAt     DateTime  @updatedAt

-  @@map(name: "users")
 }
 
 model VerificationRequest {
-  id         Int      @default(autoincrement()) @id
+  id         Int      @id @default(autoincrement())
   identifier String
   token      String   @unique
   expires    DateTime
-  createdAt  DateTime  @default(now()) @map(name: "created_at")
+  createdAt  DateTime @default(now())
-  updatedAt  DateTime  @default(now()) @map(name: "updated_at")
+  updatedAt  DateTime @updatedAt

-  @@map(name: "verification_requests")
 }
```

### CODE REVIEW

File: packages/prisma/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+    }
+
+    async function linkAccount(
+      userId: number,
+      providerId: string,
+      providerType: string,
+      providerAccountId: string,
+      refreshToken: string,
+      accessToken: string,
+      accessTokenExpires: string | Date | null | undefined
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:225]:
```suggestion
      accessTokenExpires?: string | Date | null
```

File: packages/prisma/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+        if (cachedSession) {
+          debug("GET_SESSION - Fetched from LRU Cache", cachedSession);
+          return cachedSession;
+        }
+        const session = await prisma[Session as "session"].findUnique({
+          where: { sessionToken: sessionToken },
+        });
+
+        // Check session has not expired (do not return it if it has)
+        if (session && session.expires && new Date() > session.expires) {
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:325]:
```suggestion
        if (session?.expires && new Date() > session.expires) {
```

File: packages/prisma/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+          }
+        }
+
+        const { id, expires } = session;
+        sessionCache.set(session.sessionToken, session, maxAge(expires));
+        const updatedSession = await prisma[Session as "session"].update({
+          where: { id },
+          data: { expires },
+        });
+        return updatedSession;
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:392]:
```suggestion
        return prisma[Session as "session"].update({
          where: { id },
          data: { expires },
        })
```

File: packages/prisma/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+      }
+    }
+
+    async function deleteSession(sessionToken: string) {
+      debug("DELETE_SESSION", sessionToken);
+      try {
+        sessionCache.del(sessionToken);
+        return await prisma[Session as "session"].delete({
+          where: { sessionToken },
+        });
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:406]:
```suggestion
        return prisma[Session as "session"].delete({
          where: { sessionToken },
        });
```

File: packages/prisma/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+        const verificationRequest = await prisma[
+          VerificationRequest as "verificationRequest"
+        ].findUnique({
+          where: { token: hashedToken },
+        });
+
+        if (
+          verificationRequest &&
+          verificationRequest.expires &&
+          new Date() > verificationRequest.expires
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:491]:
```suggestion
          verificationRequest?.expires &&
          new Date() > verificationRequest.expires
```

File: packages/prisma/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+    ) {
+      debug("DELETE_VERIFICATION", identifier, token);
+      try {
+        // Delete verification entry so it cannot be used again
+        const hashedToken = createHash("sha256")
+          .update(`${token}${secret}`)
+          .digest("hex");
+        return await prisma[VerificationRequest as "verificationRequest"].delete({
+          where: { token: hashedToken },
+        });
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:522]:
```suggestion
        return prisma[VerificationRequest as "verificationRequest"].delete({
          where: { token: hashedToken },
        });
```

File: packages/prisma/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+      linkAccount,
+      unlinkAccount,
+      createSession,
+      getSession,
+      updateSession,
+      deleteSession,
+      createVerificationRequest,
+      getVerificationRequest,
+      deleteVerificationRequest,
+    });
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:546]:
```suggestion
    return {
      createUser,
      getUser,
      getUserByEmail,
      getUserByProviderAccountId,
      updateUser,
      deleteUser,
      linkAccount,
      unlinkAccount,
      createSession,
      getSession,
      updateSession,
      deleteSession,
      createVerificationRequest,
      getVerificationRequest,
      deleteVerificationRequest,
    }
```

File: packages/prisma/src/index.ts (hunk 3 above)
Comment (balazsorban44) [inline packages/prisma/src/index.ts:392]:
```suggestion
        return prisma[Session as "session"].update({
          where: { id },
          data: { expires },
        })
```

File: packages/prisma/src/index.ts (hunk 2 above)
Comment (balazsorban44) [inline packages/prisma/src/index.ts:325]:
```suggestion
        if (session?.expires && new Date() > session.expires) {
```

File: packages/prisma/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+          data: {
+            accessToken,
+            refreshToken,
+            providerAccountId: `${providerAccountId}`,
+            providerId,
+            providerType,
+            accessTokenExpires,
+            user: { connect: { id: userId } },
+          },
+        });
```
Comment (balazsorban44) [inline packages/prisma/src/index.ts:248]:
```suggestion
        return prisma[Account as "account"].create({
          data: {
            accessToken,
            refreshToken,
            providerAccountId: `${providerAccountId}`,
            providerId,
            providerType,
            accessTokenExpires,
            user: { connect: { id: userId } },
          },
        });
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Just some formatting suggestions, looks good otherwise! Thanks!

File: packages/prisma/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+      );
+    }
+
+    const defaultSessionMaxAge = 30 * 24 * 60 * 60 * 1000;
+    const sessionMaxAge = appOptions?.session?.maxAge
+      ? appOptions.session.maxAge * 1000
+      : defaultSessionMaxAge;
+    const sessionUpdateAge = appOptions?.session?.updateAge
+      ? appOptions.session.updateAge * 1000
+      : 0;
```
[the remark this answers is not in the document]
Comment (williamluke4) [reply packages/prisma/src/index.ts:93]: Here shouldn't the  `defaultSessionMaxAge` && `appOptions?.session?.maxAge` remain as ms, not seconds.

**Review by balazsorban44 — APPROVED** [review]
LGTM!

**Pull request merged**